### PR TITLE
Align .tbss section to 8-byte boundary

### PR DESCRIPTION
**Pull request by ericlove**
When compiling with `-ffunction-sections -fdata-sections`, the removal of
unused functions sometimes results in the misalignment of other sections,
causing traps.  This fixes the problem for at least `.tbss`

### CODE REVIEW

File: templates/base.lds [hunk 1]
```diff
@@ -259,7 +259,7 @@ SECTIONS
     PROVIDE( metal_segment_data_target_start = ADDR(.data) );
     PROVIDE( metal_segment_data_target_end = ADDR(.tdata) + SIZEOF(.tdata) );
 
-    .tbss : {
+    .tbss : ALIGN(8) {
```
Comment (keith-packard) [inline templates/base.lds:262]: I think we need ALIGN(8) on the .tbss_space section as well, to make sure the space allocated for tbss is the same as the amount we used.

Comment (ericlove) [reply templates/base.lds:262]: I've made another commit to align .tbss_space as well.  What about .ctors, .dtors, .text, .tdata, and .rodata?

**Comment by knightsifive** [comment]
I don't fully understand the bug that this fixes, or how it fixes it, and I'm concerned that there's an underlying problem that's not being addressed. LD (obeying the System V gABI) should guarantee that the alignment of an output section equals the strictest alignment of its input sections. I don't see why this guarantee should be violated if the linker decides to garbage-collect. So I'm concerned the compiler/assembler is not emitting the appropriate alignment directives somewhere, and we've just been getting lucky.

More generally I'm concerned by the haphazard sprinkling of `ALIGN`s in our link scripts: many of them seem unnecessary to me.

**Comment by ericlove** [comment]
Yeah, it seems like `.tbss` sections emitted in each object file should be aligned to whatever the alignment of the stack is.  I wonder whether perhaps a `#define ASM_OUTPUT_ALIGNED_BSS` is required in either [gcc/config/riscv/riscv.h](https://github.com/riscv/riscv-gcc/blob/6b10a5547b5476ee3b6bd59af4d86600395c008a/gcc/config/riscv/riscv.h) or [gcc/config/riscv/elf.h](https://github.com/riscv/riscv-gcc/blob/6b10a5547b5476ee3b6bd59af4d86600395c008a/gcc/config/riscv/elf.h) ?  This is used in [varasm.c](https://github.com/riscv/riscv-gcc/blob/6b10a5547b5476ee3b6bd59af4d86600395c008a/gcc/varasm.c#L2034), it seems, but I don't really know my way around the GCC source tree well enough to tell how this alignment is actually supposed to get generated by the compiler (or how tbss is associated with bss)...

**Comment by knightsifive** [comment]
@jim-wilson do you happen to know if/how RISC-V GCC emits alignment directives for `.bss` and `.tbss` sections?

**Comment by jim-wilson** [comment]
The compiler won't align the bss or tbss sections.  It will only align objects inside those sections.  The linker will align the sections to the max alignment of the objects inside the section.  The linker script may optionally align the sections, which has no effect unless it is larger than the max object alignment.  The linker script can be supplied by the user, so this part is unpredictable.  The default linker script for riscv isn't aligning either section.  Though curiously it is trying to align the end of the .bss section, apparently to avoid padding between the end of .bss and the _end symbol.

**Comment by keith-packard** [comment]
.tdata and .tbss are handled by 'magic' in the linker, and so perhaps alignment doesn't apply to them like it would with other sections. Let's add ALIGN(8) to .tdata, .tbss and .tbss_space while we investigate the need for ALIGN on other sections.

File: templates/base.lds (hunk 1 above)
Comment (keith-packard) [reply templates/base.lds:262]: As others have mentioned, those *should* all get automatically aligned by the linker.

Comment (keith-packard) [reply templates/base.lds:262]: We also need to align .tdata -- if .data doesn't end on an aligned address, then all of the TLS data will end up unaligned, and that's bad...

**Pull request merged**